feat(changelog): Add support for azure changelogs

larsf96 opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->

Adds support for Azure Changelogs. Suggested in PR
- #37745

## Context

Please select one of the below:

- [x] This closes an existing Issue: #18324
- [ ] This doesn't close an Issue, but I accept the risk that this PR may be closed if maintainers disagree with its opening or implementation

## AI assistance disclosure

<!-- We request this information to assist reviewers in identifying AI-generated errors and other issues specific to AI usage. While we typically permit the use of AI tools, we appreciate being notified when they are employed. -->

Did you use AI tools to create any part of this pull request?

Please select one option and, if yes, briefly describe how AI was used (e.g., code, tests, docs) and which tool(s) you used.

- [ ] No — I did not use AI for this contribution.
- [ ] Yes — minimal assistance (e.g., IDE autocomplete, small code completions, grammar fixes).
- [x] Yes — substantive assistance (AI generated non‑trivial portions of code, tests, or documentation).
- [ ] Yes — other (please describe):

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

The public repository: <URL>

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

larsf96 committed:
feat(changelog): Add support for azure changelogs

larsf96 committed:
feat(changelog): Add support for azure changelogs

larsf96 committed:
refactor: Use node azure devops api instead of HTTP wrapper

larsf96 committed:
refactor: Use node azure devops api instead of HTTP wrapper

larsf96 committed:
fix: refactored tests

larsf96 committed:
fix: refactored tests

larsf96 committed:
fix tests

larsf96 committed:
fix tests

larsf96 committed:
fix test

larsf96 committed:
fix test

larsf96 committed:
Add more tests

larsf96 committed:
Add more tests

larsf96 committed:
added more tests

larsf96 committed:
added more tests

larsf96 committed:
fix lintint + tests

larsf96 committed:
fix lintint + tests

larsf96 committed:
Add more tests

larsf96 committed:
Add more tests

larsf96 commented on lib/workers/repository/update/pr/changelog/release-notes.spec.ts:
```diff
@@ -162,6 +186,12 @@
+const azureProject = partial<ChangeLogProject>({
+  type: 'azure',
+  apiBaseUrl: 'https://dev.azure.com/some-org/some-project/_apis/',
+  baseUrl: 'https://dev.azure.com/some-org/some-project/',
```
I would still need some advisory on this one. The tests will fail if I escape the dots

illay1994 replied:
Hi @larsf96 

I spent half a day yesterday and a few hours today already.

I can't find a way to fix it.

Perhaps we could solicit input from the maintainers.

larsf96 replied:
Hi @illay1994 
I think that would be good, maybe it's just a false positive..

larsf96 replied:
Hi @viceice 
Do you have some guidance here please?

illay1994 replied:
Hi @larsf96 

I found a fix in the scope of PR #44130, I tested it.

larsf96 committed:
add test

larsf96 committed:
add test

larsf96 committed:
Add coverage exception

larsf96 committed:
Add coverage exception

larsf96 committed:
reverted snapshots

larsf96 committed:
reverted snapshots

larsf96 committed:
refactored some parts

larsf96 committed:
refactored some parts

larsf96 committed:
Update lib/modules/platform/azure/azure-helper.ts

Co-authored-by: Michael Kriese <michael.kriese@gmx.de>

larsf96 committed:
Update lib/modules/platform/azure/azure-helper.ts

Co-authored-by: Michael Kriese <michael.kriese@gmx.de>

larsf96 committed:
removed beforeEach

larsf96 committed:
removed beforeEach

larsf96 committed:
fix test

larsf96 committed:
fix test

larsf96 committed:
add project to Azure calls (required by Azure Devops (server))

larsf96 committed:
add project to Azure calls (required by Azure Devops (server))

larsf96 committed:
Add tests for project extraction

larsf96 committed:
Add tests for project extraction

larsf96 committed:
refactor: Refactored logic to only modify logic for platform azure

larsf96 committed:
refactor: Refactored logic to only modify logic for platform azure

larsf96 committed:
fixed CodeQL comments (credits @illay1994)

larsf96 committed:
fixed CodeQL comments (credits @illay1994)

larsf96 committed:
Merge branch 'main' into feature/azure-changelog

larsf96 committed:
fix lint issues

larsf96 committed:
fix lint issues

larsf96 committed:
fix imports

larsf96 committed:
fix imports

larsf96 committed:
fix import

larsf96 committed:
fix import

larsf96 committed:
refactored logic

larsf96 committed:
refactored logic

larsf96 committed:
Merge branch 'main' into feature/azure-changelog

larsf96 committed:
Add tests

larsf96 committed:
Add tests

larsf96 committed:
Merge branch 'feature/azure-changelog' of https://github.com/larsf96/renovate into feature/azure-changelog

larsf96 committed:
add test

larsf96 committed:
add test

larsf96 committed:
Merge branch 'feature/azure-changelog' of https://github.com/larsf96/renovate into feature/azure-changelog

larsf96 committed:
Merge branch 'main' into feature/azure-changelog

larsf96 committed:
Merge branch 'feature/azure-changelog' of https://github.com/larsf96/renovate into feature/azure-changelog

larsf96 committed:
lint: fix linting

illay1994 commented on lib/workers/repository/update/pr/changelog/release-notes.ts:
```diff
@@ -91,22 +94,23 @@ export function massageBody(
   input: string | undefined | null,
   baseUrl: string,
 ): string {
+  const baseUrlForRegex = baseUrl.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
```
```suggestion
  const baseUrlForRegex = RegExp.escape(baseUrl);
```

larsf96 committed:
Update lib/workers/repository/update/pr/changelog/release-notes.ts

Co-authored-by: Illia Mazuriak <illay1994@gmail.com>

larsf96 committed:
Merge branch 'main' into feature/azure-changelog

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/release-notes.ts:
```diff
@@ -389,11 +400,22 @@ export async function getReleaseNotesMd(
+                  parenthesizedHeading
+                    .replace(regEx(/^\s*#*\s*/), '')
+                    .toLowerCase()
+                    .replace(regEx(/\s+/), '-'),
```
use a global flag so all - gets replaced

RahulGautamSingh replied:
the regex has no `g` flag, so only the first whitespace run is replaced. a multi-word heading keeps its later spaces and produces a broken anchor. use `/\s+/g`.

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/types.ts:
```diff
@@ -43,6 +44,7 @@ export interface ChangeLogProject {
 }
 
 export type ChangeLogError =
+  | 'MissingAzureToken'
```
this is not use anywhere though

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/azure/index.ts:
```diff
@@ -0,0 +1,90 @@
+    .sort((a, b) => compareChangelogFilePath(a.relativePath!, b.relativePath!))
+    .shift()?.relativePath;
+
+  /* v8 ignore next -- not testable */
```
The : '' branch is testable (call with an empty/undefined sourceDirectory) .. please add a test:

RahulGautamSingh replied:
this line runs in the tests, so the ignore reason is inaccurate. only the empty-string branch is uncovered, and it is testable by passing `sourceDirectory: ''`. add that test instead of ignoring.

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/azure/index.ts:
```diff
@@ -0,0 +1,90 @@
+
+  const tree = await azureHelper.getTrees(
+    urlEncodedRepo,
+    sourceDirectoryId.objectId!,
```
Prefer an `is` guard over !

RahulGautamSingh replied:
the `objectId!` non-null assertion bypasses the type system. guard the value and return null if it is missing, like the other fields here.

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/source.ts:
```diff
@@ -299,6 +304,11 @@ export abstract class ChangeLogSource {
+    if (this.platform === 'azure') {
+      return repositoryLength === 1;
+    } else {
+      return repositoryLength === 2;
```
drop the else (or use a ternary)  .. keep indentation low

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/azure/index.ts:
```diff
@@ -0,0 +1,90 @@
+  _release: ChangeLogRelease,
+): ChangeLogNotes[] {
+  logger.trace('azure.getReleaseList()');
+  logger.debug('Unsupported Azure DevOps feature.  Skipping release fetching.');
```
contains double spaces

RahulGautamSingh replied:
double space between `feature.` and `Skipping`.

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/azure/index.spec.ts:
```diff
@@ -0,0 +1,598 @@
+    it('handles release notes', async () => {
+      const changelogMd = Fixtures.get('jest.md', '..');
+
+      vi.spyOn(azureHelper, 'getItem').mockRejectedValue({
```
use .once here & at all other similar instances where you are doing mockRejectedValue first and mockResolvedValue second for the same fn, 

plain variants of mockRejectedValue and mockResolvedValue override the mockImplementation internally so the second call overwrites the first one (which is not what you want). `once` queues them so that is fine.

RahulGautamSingh replied:
this `mockRejectedValue` is overwritten by the later `mockResolvedValue` on the same `getItem`, so it never runs. it also rejects with a plain object, not an `Error`. remove the dead reject, or use `mockResolvedValueOnce` if sequencing was intended. same pattern repeats in the tests below.

illay1994 committed:
Fixed comments

RahulGautamSingh commented on lib/modules/platform/azure/azure-helper.ts:
```diff
@@ -220,3 +222,43 @@ export async function getAllProjectTeams(
   return allTeams;
 }
+
+export async function getTags(repoId: string): Promise<GitRef[]> {
```
`getTags` is never called. Changelog tags come through the base `getAllTags` and the `azure-tags` datasource. Remove this function and its test.

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/azure/index.ts:
```diff
@@ -0,0 +1,90 @@
+  const urlEncodedRepo = encodeURIComponent(repository);
+
+  // Extract project name from API base URL (last path segment before "_apis/")
+  const projectMatch = /\/([^/]+)\/_apis\//.exec(apiBaseUrl);
```
use the `regEx()` util from `util/regex.ts` instead of a raw literal, and use a named group like `(?<project>[^/]+)`. positional `match[1]` breaks silently if the pattern changes.

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/azure/index.ts:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  let changelogFile = files
+    .sort((a, b) => compareChangelogFilePath(a.relativePath!, b.relativePath!))
```
`relativePath!` bypasses the type system. `relativePath` is optional on `GitTreeEntryRef`, so guard it rather than assert.

RahulGautamSingh reviewed (commented):
Review of the Azure changelog support. A few correctness items and some test cleanup. Comments inline.


_AI used_

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/azure/index.spec.ts:
```diff
@@ -0,0 +1,598 @@
+      expect(res).toBeNull();
+    });
+
+    it('handles null tree entries', async () => {
```
this test is identical to `handles empty tree entries`, both set `treeEntries: []`. the comment says null entries but none are set. drop one or make them differ.

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/azure/index.spec.ts:
```diff
@@ -0,0 +1,598 @@
+      );
+    });
+
+    it('extracts project name with special characters', async () => {
```
same url and assertions as the previous test. consolidate the two.

RahulGautamSingh commented on lib/workers/repository/update/pr/changelog/source.ts:
```diff
@@ -299,6 +304,11 @@ export abstract class ChangeLogSource {
   }
 
   hasValidRepository(repository: string): boolean {
```
the `else` after a `return` is unneeded. return early for the azure case and drop the else to keep nesting low.

larsf96 committed:
Fix comments

larsf96 committed:
Merge branch 'main' into feature/azure-changelog

RahulGautamSingh reviewed (approved):
lgtm

larsf96 committed:
Merge branch 'main' into feature/azure-changelog

larsf96 committed:
fix duplicate hostRules import

viceice commented on lib/workers/repository/update/pr/changelog/azure/index.ts:
```diff
@@ -0,0 +1,98 @@
+
+  // Extract project name from API base URL (last path segment before "_apis/")
+  const projectMatch = regEx('/(?<project>[^/]+)/_apis/').exec(apiBaseUrl);
+  const project = projectMatch?.groups?.project ?? '';
```
use `coerceString`

viceice commented on lib/workers/repository/update/pr/changelog/azure/index.ts:
```diff
@@ -0,0 +1,98 @@
+export async function getReleaseNotesMd(
+  repository: string,
+  apiBaseUrl: string,
+  sourceDirectory = '/',
```
renovate could pass `null` here due to reasons, so you should use `coerceString(sourceDirectory, '/')`

viceice commented on lib/workers/repository/update/pr/changelog/azure/index.ts:
```diff
@@ -0,0 +1,98 @@
+
+  let changelogFile = files
+    .sort((a, b) => compareChangelogFilePath(a.relativePath, b.relativePath))
+    .shift()?.relativePath;
```
```suggestion
    .shift()!.relativePath;
```

we know there is at least one item

viceice commented on lib/workers/repository/update/pr/changelog/azure/index.ts:
```diff
@@ -0,0 +1,98 @@
+    .sort((a, b) => compareChangelogFilePath(a.relativePath, b.relativePath))
+    .shift()?.relativePath;
+
+  changelogFile = `${sourceDirectory ? ensureTrailingSlash(sourceDirectory) : ''}${changelogFile}`;
```
use coerced `sourceDirectory `.

is `relativePath` really releative to the original passed `sourceDirectory `?

larsf96 replied:
yes, just tested that again with a real Azure Devops instance, it's the relative path

viceice commented on lib/workers/repository/update/pr/changelog/azure/source.ts:
```diff
@@ -0,0 +1,41 @@
+    prevHead: string,
+    nextHead: string,
+  ): string {
+    const regex = regEx(`^refs/tags/`, undefined, false);
```
why no caching?

viceice commented on lib/workers/repository/update/pr/changelog/azure/source.ts:
```diff
@@ -0,0 +1,41 @@
+    }
+    const protocol = parsedUrl.protocol;
+    const host = parsedUrl.host;
+    const pathname = parsedUrl.pathname.slice(1).split('/');
+    const organization = pathname[0];
+    const projectName = pathname[1];
```
```suggestion
    const [ organization, projectName ] = parsedUrl.pathname.slice(1).split('/');
```

simplify

viceice commented on lib/workers/repository/update/pr/changelog/release-notes.spec.ts:
```diff
@@ -1685,6 +1751,38 @@ describe('workers/repository/update/pr/changelog/release-notes', () => {
     });
 
+    it('handles azure sourceDirectory', async () => {
+      const sourceDirectory = '/packages/foo';
```
are we sure `sourceDirectory` always starts with a slash?

larsf96 replied:
Yes, at least the Azure API is adding that in the response even if you request the path without it

viceice commented on lib/workers/repository/update/pr/changelog/release-notes.ts:
```diff
@@ -392,11 +403,22 @@ export async function getReleaseNotesMd(
-            .join('-')
-            .replace(regEx(/[^A-Za-z0-9-]/g), '');
-          const url = `${notesSourceUrl}#${mdHeadingLink}`;
+          const mdHeadingLink =
+            project.type === 'azure'
+              ? encodeURIComponent(
+                  parenthesizedHeading
+                    .replace(regEx(/^\s*#*\s*/), '')
+                    .toLowerCase()
+                    .replace(regEx(/\s+/g), '-'),
+                )
+              : title
+                  .filter((word) => !isHttpUrl(word))
+                  .join('-')
+                  .replace(regEx(/[^A-Za-z0-9-]/g), '');
+          const url =
+            project.type === 'azure'
+              ? `${notesSourceUrl}&anchor=${mdHeadingLink}`
+              : `${notesSourceUrl}#${mdHeadingLink}`;
```
should be a new function on `ChangeLogSource` with original implementation and then the Azure source can override.
that makes this code much more readable.
should be done as prefactor PR

viceice commented on lib/workers/repository/update/pr/changelog/release-notes.ts:
```diff
@@ -552,6 +574,10 @@ function getNotesSourceUrl(
     );
   }
 
+  if (project.type === 'azure') {
```
same here

viceice commented on lib/workers/repository/update/pr/changelog/source.ts:
```diff
@@ -287,6 +289,9 @@ export abstract class ChangeLogSource {
       return '';
     }
     const pathname = parsedUrl.pathname;
+    if (this.platform === 'azure') {
```
you should override `getRepositoryFromUrlfunctions` and `hasValidRepository` in `AzureChangeLogSource` instead of doing conditionals here.
makes code much more readable.

viceice commented on lib/workers/repository/update/pr/changelog/azure/index.spec.ts:
```diff
@@ -0,0 +1,587 @@
+          ['https://dev.azure.com/org/project/_git/repo/', 'repo'],
+        ];
+
+        for (const [sourceUrl, expectedRepo] of testCases) {
```
use `it.each`

larsf96 committed:
enable cache

larsf96 committed:
use coerced sourceDirectory

larsf96 committed:
add test for input sourceDirectory without leading slash

larsf96 committed:
use it.each

larsf96 committed:
Merge remote-tracking branch 'upstream/main' into feature/azure-changelog

# Conflicts:
#	lib/workers/repository/update/pr/changelog/release-notes.ts

larsf96 committed:
refactor(changelog): move azure URL and repo overrides into AzureChangeLogSource

Follows up prefactor #45260. Overrides getNotesSourceUrl,
getReleaseNotesMdAnchorUrl, getRepositoryFromUrl and hasValidRepository
in AzureChangeLogSource instead of branching on project.type/platform in
the base ChangeLogSource and release-notes. Simplify getBaseUrl
destructuring and add a null-URL test for getRepositoryFromUrl.

larsf96 commented:
@viceice I refactored the overrides after the prefactor PR was merged. Hope that everything is fine now
